### PR TITLE
state: initial version for the consul state store implementation

### PR DESCRIPTION
**Pull request by anubhavmishra**
# Description

This is an initial version of HashiCorp [Consul](https://consul.io) state store implementation.

## Issue reference

The PR closes: #5 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [ ] Extended the documentation

### CODE REVIEW

**Comment by anubhavmishra** [comment]
Once this PR is merged, we can review: https://github.com/dapr/dapr/pull/686

**Comment by anubhavmishra** [comment]
@yaron2 that's a lot of folks on review. 😄 Do you think this will be reviewed soon?

File: state/consul/consul.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+		Value: reqValByte}, nil)
+
+	if err != nil {
+		return fmt.Errorf("couldn't set key %s: %s", keyWithPath, err)
+	}
+
+	return nil
+}
+
+// BulkSet performs a bulks save operation
```
Comment (lkysow) [inline state/consul/consul.go]:
```suggestion
// BulkSet performs a bulk save operation
```

**Comment by yaron2** [comment]
> @yaron2 that's a lot of folks on review. 😄 Do you think this will be reviewed soon?

Today!

File: state/consul/consul.go [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+	b, ok := req.Value.([]byte)
+	if ok {
+		reqValByte = b
+	} else {
+		reqValByte, _ = json.Marshal(req.Value)
+	}
+
+	keyWithPath := fmt.Sprintf("%s/%s", c.keyPrefixPath, req.Key)
+
+	_, err := c.client.KV().Put(&api.KVPair{
```
Comment (yaron2) [inline state/consul/consul.go]: Does the `Put` operation also take in a consistency model? (strong, for example)

File: state/consul/consul.go [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+	client, err := api.NewClient(config)
+	if err != nil {
+		return errors.Wrap(err, "initializing consul client")
+	}
+
+	var queryOpts *api.QueryOptions
+
+	c.client = client
+	c.queryOpts = queryOpts
```
[the remark this answers is not in the document]
Comment (anubhavmishra) [reply state/consul/consul.go]: yea, it is. Nice catch!

File: state/consul/consul.go [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+	"encoding/json"
+	"fmt"
+
+	"github.com/dapr/components-contrib/state"
+	"github.com/hashicorp/consul/api"
+	"github.com/pkg/errors"
+)
+
+// StateStore is a state store implementation for HashiCorp Consul.
+type StateStore struct {
```
[the remark this answers is not in the document]
Comment (anubhavmishra) [reply state/consul/consul.go]: For sure.

File: state/consul/consul.go (hunk 2 above)
Comment (anubhavmishra) [reply state/consul/consul.go]: As far as I know, the `Put` operation currently doesn't support the consistency model.

https://www.consul.io/api/kv.html#create-update-key

**Comment by yaron2** [comment]
One last thing: update the vendor dir with `go mod vendor`

File: state/consul/consul.go (hunk 2 above)
Comment (yaron2) [reply state/consul/consul.go]: ok

**Comment by anubhavmishra** [comment]
@yaron2 we should be good to go! Thank you so much for a swift and awesome review.

File: state/consul/consul.go [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+	"github.com/dapr/components-contrib/state"
+	"github.com/hashicorp/consul/api"
+	"github.com/pkg/errors"
+)
+
+// Consul is a state store implementation for HashiCorp Consul.
+type Consul struct {
+	client        *api.Client
+	queryOpts     *api.QueryOptions
```
Comment (yaron2) [inline state/consul/consul.go]: this is not used.. can you remove it?

File: state/consul/consul.go [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	return &config, nil
+}
+
+// Get retrieves a Consul KV item
+func (c *Consul) Get(req *state.GetRequest) (*state.GetResponse, error) {
+	var queryOpts *api.QueryOptions
```
Comment (yaron2) [inline state/consul/consul.go]: Can this turn to `queryOpts := &api.QueryOptions{}`?

File: state/consul/consul.go [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+	if req.Options.Consistency == state.Strong {
+		queryOpts.RequireConsistent = true
+	}
+
+	resp, queryMeta, err := c.client.KV().Get(fmt.Sprintf("%s/%s", c.keyPrefixPath, req.Key), c.queryOpts)
```
Comment (yaron2) [inline state/consul/consul.go]: Use the `queryOpts` variable from above instead of `c.queryOpts`

**Comment by yaron2** [comment]
> @yaron2 we should be good to go! Thank you so much for a swift and awesome review.

Almost there.. :)

**Comment by yaron2** [comment]
LGTM

**Comment by amanbha** [comment]
This PR closed #51

**Pull request merged**